### ISSUE TITLE
Why do we need to write to local first, then calculate IOU and format result when doing test.

### ISSUE THREAD

**Issue by jianlong-yuan** [open-mmlab/mmsegmentation#395]
![image](https://user-images.githubusercontent.com/18137551/109602470-1df88a00-7b5b-11eb-9e5a-7c5afb8d0ec0.png)

This will not only increase the storage during the test but also increase the memory of the GPU

**Comment by xvjiarui** [comment]
@jianlong-yuan 
We support two collect method. However, for segmentation, gpu_collect is not feasible since the entire dataset won't fit in GPU.

**Comment by jianlong-yuan** [comment]
Cpu collect method will cost much storage memory.  When we have a lot of test images, it is not suitable.

**Comment by xvjiarui** [comment]
Hi @jianlong-yuan 
You may try `efficient_test` options when testing.

**Comment by jianlong-yuan** [comment]
![image](https://user-images.githubusercontent.com/18137551/110083775-15f25180-7dca-11eb-80cc-a7e54b3009bc.png)

![image](https://user-images.githubusercontent.com/18137551/110084022-6bc6f980-7dca-11eb-8294-5bdd0b031ac8.png)

I have tried. As shown in Figure, we must save outputs to temp files. This is much cost!

**Comment by xvjiarui** [comment]
Hi @jianlong-yuan 
If we don't save them into filesystem, it will cost a lot of memory, which is also not ideal.

**Comment by jianlong-yuan** [comment]
> Hi @jianlong-yuan
> If we don't save them into filesystem, it will cost a lot of memory, which is also not ideal.

Actually, this is more efficient, https://github.com/hszhao/semseg/blob/master/tool/test.py and https://github.com/LikeLy-Journey/SegmenTron/blob/master/tools/eval.py

**Comment by ywwwer** [comment]
> > Hi @jianlong-yuan
> > If we don't save them into filesystem, it will cost a lot of memory, which is also not ideal.
> 
> Actually, this is more efficient, https://github.com/hszhao/semseg/blob/master/tool/test.py and https://github.com/LikeLy-Journey/SegmenTron/blob/master/tools/eval.py

Agreed. If it's the prediction results are not needed,  it's way more efficient to compare gt and pred immediately in each batch than saving the results in disk/memory and then loading back.

**Comment by xvjiarui** [comment]
Hi @ywwwer 
It is kind of tricky. Since we decouple the test and evaluation. What test phase should return is the output results. The evaluation phase will evaluate these outputs. We may consider refactoring this part in the future.

**Comment by jianlong-yuan** [comment]
@xvjiarui  But if our disk space is insufficient, or the disk io is low, it will seriously affect the efficiency of the test.

**Comment by xvjiarui** [comment]
@jianlong-yuan 
That's one point. We will consider about it.

**Comment by jianlong-yuan** [comment]
Any news?
